packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmValueAllocator.kt

```diff
@@ -0,0 +1,184 @@
+        null -> nullTransport()
+        else -> createTransport(realm_value_type.RLM_TYPE_FLOAT) {
+            fnum = value
+        }
+    }
+
+    override fun doubleTransport(value: Double?): RealmValue = when (value) {
+        null -> nullTransport()
+        else -> createTransport(realm_value_type.RLM_TYPE_DOUBLE) {
+            dnum = value
+        }
+    }
+
+    override fun objectIdTransport(value: ByteArray?): RealmValue = when (value) {
+        null -> nullTransport()
+        else -> createTransport(realm_value_type.RLM_TYPE_OBJECT_ID) {
+            object_id.apply {
+                // TODO BENCHMARK: is this faster than memcpy? see 'uuidTransport'
+                (0 until OBJECT_ID_BYTES_SIZE).map {
+                    bytes[it] = value[it].toUByte()
+                }
+            }
+        }
+    }
+
+    override fun uuidTransport(value: ByteArray?): RealmValue = when (value) {
+        null -> nullTransport()
+        else -> createTransport(realm_value_type.RLM_TYPE_UUID) {
+            uuid.apply {
+                // TODO BENCHMARK: is this faster than looping through the structure? see 'objectIdTransport'
+                value.usePinned {
+                    val dest = bytes.getPointer(scope)
+                    val source = it.addressOf(0)
+                    memcpy(dest, source, UUID_BYTES_SIZE.toULong())
+                }
+            }
+        }
+    }
+
+    override fun realmObjectTransport(value: RealmObjectInterop?): RealmValue = when (value) {
+        null -> nullTransport()
+        else -> createTransport(realm_value_type.RLM_TYPE_LINK) {
+            realm_wrapper.realm_object_as_link(value.objectPointer.cptr())
+                .useContents {
+                    link.apply {
+                        target_table = this@useContents.target_table
+                        target = this@useContents.target
+                    }
+                }
+        }
+    }
+
+    override fun stringTransport(value: String?): RealmValue = when (value) {
+        null -> nullTransport()
+        else -> createTransport(realm_value_type.RLM_TYPE_STRING) {
+            string.set(scope, value)
+        }
+    }
+
+    override fun byteArrayTransport(value: ByteArray?): RealmValue = when (value) {
+        null -> nullTransport()
+        else -> createTransport(realm_value_type.RLM_TYPE_BINARY) {
+            binary.set(scope, value)
+        }
+    }
+
+    override fun queryArgsOf(queryArgs: Array<RealmValue>): RealmQueryArgsTransport {
+        val cArgs = scope.allocArray<realm_query_arg_t>(queryArgs.size)
+        queryArgs.mapIndexed { i, arg ->
+            cArgs[i].apply {
+                this.nb_args = 1.toULong()
+                this.is_list = false
+                this.arg = arg.value.ptr
+            }
+        }
+        return RealmQueryArgsTransport(cArgs.pointed)
+    }
+
+    override fun free() {
+        scope.clear()
+    }
+
+    private inline fun createTransport(
```

Comment:
See common on JVM implementation on suggestion to handle `null` here instead of all the individual switches